### DISCUSSION TITLE
Enabling Overlay for ltc294x.dtbo battery gauge

### DISCUSSION THREAD

**Discussion by pterodyne** [Joshua-Riek/ubuntu-rockchip#1000] [Q&A]
Hello all.  I'm a bit unschooled in this area, but have good experience coming from the pi ecosystem.  I'm [building a laptop](https://github.com/pterodyne/picore/) around a Radxa CM5/CM4 nano carrier, and so far everything is working well with this build (6.1 Ubuntu 24.04). I'm using [Decktility ](https://github.com/ByteWelder/Decktility/) project as the basis for the power control and monitoring.  In a nutshell it uses an Arduino to simulate a ltc294x chip, which is [natively supported](https://github.com/torvalds/linux/blob/master/drivers/power/supply/ltc2941-battery-gauge.c) by Linux.  On a pi it works great, and other than enabling the overlay, it reports the battery back to the system without doing anything else.  Anyway, Id hoped it would be as easy as enabling the dtbo, but it doesn't exist in the device tree.

I've looked all over on Radxa's site other GitHub etc.  So I went as far as de-compiling the RPI one and removing the compatibility reference to the Broadcom.  Side note none of the dtbo files I de-compiled from Rockchip had this compatibility reference so I just removed the lines and recompiled, added to u-boot, and ran uboot-update.  I can confirm that the extlinux.conf is updated with the reference, but it doesn't load.

Other considerations.  I don't completely understand the GPIO workings on the Rockchip, but it's wired in to pins 3,5 for I2C on the Pi so I'm at least guessing that it will need to have those pins enabled via dtbo as well.  Ive never dealt with compiling kernel drivers before so a bit at a loss here.  Maybe I need to compile the kernel driver as well as the dtbo.

Anyway thanks for looking and any advise!

Bryan

**Comment by Joshua-Riek** [comment]
The overlay will likely need some changes, but first, have you confirmed the kernel driver for this device is enabled?

**Reply by pterodyne** [reply]
Thanks for replying.  I have yet to figure out how.  But I think not.  Ill see if I can figure that out.

**Reply by pterodyne** [reply]
This looks promising...

```
ls -R /lib/modules/$(uname -r)/kernel/|grep ltc
ltc2945.ko
ltc2947-core.ko
ltc2947-i2c.ko
ltc2947-spi.ko
ltc2990.ko
ltc4151.ko
ltc4215.ko
ltc4222.ko
ltc4245.ko
ltc4260.ko
ltc4261.ko
i2c-mux-ltc4306.ko
ltc2983.ko
```

here is lsmod:

```
lsmod
Module                  Size  Used by
joydev                 24576  0
hid_multitouch         24576  0
xt_MASQUERADE          16384  2
xt_mark                16384  2
nft_compat             20480  4
nft_chain_nat          16384  2
nf_nat                 69632  2 nft_chain_nat,xt_MASQUERADE
nf_conntrack          184320  2 nf_nat,xt_MASQUERADE
nf_defrag_ipv6         24576  1 nf_conntrack
nf_defrag_ipv4         16384  1 nf_conntrack
tun                    49152  2
nf_tables             245760  99 nft_compat,nft_chain_nat
qrtr                   45056  4
squashfs               61440  5
r8153_ecm              16384  0
input_leds             16384  0
88XXau               2269184  0
r8152                  98304  1 r8153_ecm
binfmt_misc            20480  1
panfrost               69632  0
drm_shmem_helper       28672  1 panfrost
gpu_sched              45056  1 panfrost
uio_pdrv_genirq        16384  0
uio                    24576  1 uio_pdrv_genirq
fuse                  135168  1
dm_mod                143360  0
nfnetlink              24576  3 nft_compat,nf_tables
ip_tables              32768  0
ipv6                  548864  62
```

**Reply by pterodyne** [reply]
after running modprobe ltc2947-i2c:
lsmod

```
Module                  Size  Used by
ltc2947_i2c            16384  0
ltc2947_core           20480  1 ltc2947_i2c
xt_MASQUERADE          16384  2
xt_mark                16384  2
nft_compat             20480  4
nft_chain_nat          16384  2
nf_nat                 69632  2 nft_chain_nat,xt_MASQUERADE
nf_conntrack          184320  2 nf_nat,xt_MASQUERADE
nf_defrag_ipv6         24576  1 nf_conntrack
nf_defrag_ipv4         16384  1 nf_conntrack
tun                    49152  2
nf_tables             245760  99 nft_compat,nft_chain_nat
qrtr                   45056  4
squashfs               61440  5
binfmt_misc            20480  1
88XXau               2269184  0
panfrost               69632  0
drm_shmem_helper       28672  1 panfrost
gpu_sched              45056  1 panfrost
input_leds             16384  0
uio_pdrv_genirq        16384  0
uio                    24576  1 uio_pdrv_genirq
dm_mod                143360  0
fuse                  135168  1
nfnetlink              24576  3 nft_compat,nf_tables
ip_tables              32768  0
ipv6                  548864  60

```

**Reply by Joshua-Riek** [reply]
I think it may not be enabled, `CONFIG_BATTERY_GAUGE_LTC2941=y` is not in the kernel config.

The ltc2947 and ltc2945 drivers look to be sensors.

**Comment by pterodyne** [comment]
do I need to do a custom kernel config? Or more specifically could it be enabled for future builds so I don't have to rely on custom kernels from here on out?

And yes, from what I can see:

cat /boot/config-$(uname -r) | grep -i ltc2941
# CONFIG_BATTERY_GAUGE_LTC2941 is not set


Thanks.



________________________________

**Reply by Joshua-Riek** [reply]
Yeah I can enable it, a kernel update should come out maybe next week? I have a few other things id like to include in the next update.

**Comment by pterodyne** [comment]
Awesome!,  Thank you, it's a really easy way to add battery sensing to a SBC (the power management module from Decktility emulating the LTC294x chips).

Do you know which overlay I need to enable to get I2C over GPIO Pins 3/5?  They default to that on the Pi


-----Original Message-----
From: Joshua Riek ***@***.******@***.***%3e>>
Reply-To: Joshua-Riek/ubuntu-rockchip ***@***.******@***.***%3e>>
To: Joshua-Riek/ubuntu-rockchip ***@***.******@***.***%3e>>
Cc: pterodyne ***@***.******@***.***%3e>>, Author ***@***.******@***.***%3e>>
Subject: Re: [Joshua-Riek/ubuntu-rockchip] Enabling Overlay for ltc294x.dtbo battery gauge (Discussion #1000)
Date: 08/22/2024 03:28:28 PM


Yeah I can enable it, a kernel update should come out maybe next week? I have a few other things id like to include in the next update.

**Reply by Joshua-Riek** [reply]
You will likely need to make your own overlay; it depends on what GPIO pins your battery is connected to on a specific device. 

I'm very busy right now, but over the weekend I can take a look.
